src/azul/queues.py

```diff
@@ -149,66 +154,105 @@ def get_queues(self, queue_names: Iterable[str]) -> Mapping[str, Queue]:
+                'min_timeout must be less than or equal to max_timeout',
+                min_timeout, max_timeout)
+
         sleep_time = 5
-        deque_size = 10 if empty else 1
         queues = self.get_queues(config.work_queue_names)
-        queue_size_history = deque(maxlen=deque_size)
-        wait_start_time = time.time()
+        total_lengths = deque(maxlen=10 if empty else 1)
+        start_time = time.time()
+        num_bundles = 0
 
-        if not empty:
-            timeout = 2 * 60
-        elif num_bundles is None:
-            timeout = 60 * 60
-        else:
-            # calculate timeout for queues to empty with a given number of bundles
-            time_to_clear_deque = deque_size * sleep_time
-            time_processing_bundles = num_bundles * 5  # small time per bundle
-            timeout = max(time_processing_bundles + time_to_clear_deque, 60)  # at least 1 min
-            timeout = min(timeout, 60 * 60)  # at most 60 min
-
-        logger.info('Waiting up to %s seconds for %s queues to %s ...',
-                    timeout, len(queues), 'empty' if empty else 'not be empty')
         while True:
-            total_message_count = self.count_messages(queues)
-            logger.info('Counting %i total message(s) in %i queue(s).', total_message_count, len(queues))
-            queue_wait_time_elapsed = (time.time() - wait_start_time)
-            queue_size_history.append(total_message_count)
-            cumulative_queue_size = sum(queue_size_history)
-            if (cumulative_queue_size == 0) == empty and len(queue_size_history) == deque_size:
-                logger.info('The queue(s) are at the desired level.')
-                break
-            elif queue_wait_time_elapsed > timeout:
-                logger.error('The queue(s) are NOT at the desired level.')
+            # Determine queue lengths
+            total_length, queue_lengths = self._get_queue_lengths(queues)
+            total_lengths.append(total_length)
+            logger.info('Counting %i total message(s) in %i queue(s). '
+                        'The most recently sampled queue length totals are %r.',
+                        total_length, len(queue_lengths), total_lengths)
+
+            if len(total_lengths) == total_lengths.maxlen and all(n == 0 for n in total_lengths) == empty:
+                logger.info('The queues are at the desired level.')
                 break
+
+            # Determine timeout
+            if not empty:
+                timeout = 2 * 60
+            else:
+                # Estimate number of bundles first, if necessary
+                if num_expected_bundles is None:
+                    num_notifications = queue_lengths[config.notify_queue_name]
+                    if num_bundles < num_notifications:
+                        num_bundles = num_notifications
+                        start_time = time.time()  # restart the timer
```

Comment:
Agree, although some notifications are process while the timeout is increasing it should add a bit more breathing room to prevent early termination.